<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  
  
  <link rel="shortcut icon" href="../img/favicon.ico">
  <title>Scenarios - gJLS2</title>
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700" />

  <link rel="stylesheet" href="../css/theme.css" />
  <link rel="stylesheet" href="../css/theme_extra.css" />
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/styles/github.min.css" />
  
  <script>
    // Current page data
    var mkdocs_page_name = "Scenarios";
    var mkdocs_page_input_path = "UserScenarios.md";
    var mkdocs_page_url = null;
  </script>
  
  <script src="../js/jquery-2.1.1.min.js" defer></script>
  <script src="../js/modernizr-2.8.3.min.js" defer></script>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/highlight.min.js"></script>
  <script>hljs.initHighlightingOnLoad();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
        <a href=".." class="icon icon-home"> gJLS2</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="..">Home</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../Methods/">Method</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../GUI/">GUI</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../nonGUI/">nonGUI</a>
                    </li>
                </ul>
                <ul class="current">
                    <li class="toctree-l1 current"><a class="reference internal current" href="./">Scenarios</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#gjls-with-indvidual-level-genotype-and-phenotype-data">gJLS with INDVIDUAL-LEVEL genotype and phenotype data.</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#gs-with-indvidual-level-genotype-and-phenotype-data">gS with INDVIDUAL-LEVEL genotype and phenotype data.</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#gjls-with-summary-level-genotype-and-phenotype-data">gJLS with SUMMARY-LEVEL genotype and phenotype data.</a>
    </li>
    </ul>
                    </li>
                </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="..">gJLS2</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="..">Docs</a> &raquo;</li>
    
      
    
    <li>Scenarios</li>
    <li class="wy-breadcrumbs-aside">
      
    </li>
  </ul>
  
  <hr/>
</div>

          <div role="main">
            <div class="section">
              
                <h1 id="user-input-data-scenarios">User (input data) scenarios</h1>
<p>The gJSL2 analyzes each SNP at a time, it is straightforward to divide-and-conquer irrespective of your GUI choice. Here I list some possible user scenarios and our recommendation on which approach to use. We also report some performance metric to help you gauge the most appropriate option for your own computing specs.</p>
<p>To help assess the computational need, we sampled with replacement from the UKB X-chromosome data (restricted to 100 SNPs) to achieve sample sizes: 1,000, 5,000, 10,000, 50,000, 100,000, 200,000, 300,000, 400,000, and repeated the joint-location-scale analysis using <em>a single core</em> with 10GB memory via 1) PLINK R plug-in (does not support multi-thread) and 2) Rscript. The reason for keeping the 100 SNPs to estimate the performance metric is because the analysis can be easily divided to chunks (and computed over multiple cores for Rscript). To compare with X-chromosome wide analyses, we also report the memory and computational time used for UKB scale (n= 488,377, m = 15,179) for a full gJLS analysis. </p>
<p>From <strong>Figure 1</strong>, we observe roughly similar performance from the two options for sample size under 10,000, which requires ~0.30 GB and less than 30 sec of wall clock run time. This also implies that for data of this size and scale, the R-GUI options, such as Rstudio is also viable. </p>
<p>As sample size increased, PLINK becomes more demanding in terms of memory usage, but slightly faster than Rscript option. For the full UKB scale X-chromosome analysis, the computational times needed for either are very similarly: ~16 hrs for PLINK (using 3.5GB memory) and ~ 20hrs (using 1.2GB) for Rscript. This memory efficiency of Rscript is expected as the <em>BEDmatrix</em> used only map the required portion of genotype files into memory. However, the Rscript can be parallelized over multi-threads and when using 4 cores, the wall clock run time is reduced to ~10 hrs (using GB).</p>
<p><img alt="Figure 1.Performance metric for 100 SNPs from UKB" src="../figure/comp_spec.png" /></p>
<h2 id="gjls-with-indvidual-level-genotype-and-phenotype-data"><strong>gJLS</strong> with <strong>INDVIDUAL-LEVEL</strong> genotype and phenotype data.</h2>
<ul>
<li>
<p><em>I have a relatively small dataset with &lt; 5,000 samples and want to investigate particular markers or gene-set analysis (&lt; 100).</em> <span style="color:green">The R GUI should be sufficient for this purpose. PLINK binary data can be read with "BEDMatrix" and the .raw genotypes can be read in directly as a data.frame.</span></p>
</li>
<li>
<p><em>I have a relatively small dataset with &lt; 5,000 samples and want to perform a genome-wide analysis.</em> <span style="color:green">We recommend either the Rscript or PLINK R plugin as the analysis is straightforward and simple to break the job by chromosome or apply other filters within PLINK.</span></p>
</li>
<li>
<p><em>I have datasets with &gt; 5,000 samples.</em> <span style="color:green">The Rscript option will work nicely on binary files using "BEDMatrix", the multiple cores option can speed up the computational time, and the user specified write size makes sure that no results are lost in the process.</span></p>
</li>
</ul>
<h2 id="gs-with-indvidual-level-genotype-and-phenotype-data"><strong>gS</strong> with <strong>INDVIDUAL-LEVEL</strong> genotype and phenotype data.</h2>
<ul>
<li><em>I have "</em><em>INDVIDUAL-LEVEL</em><em>" genotype and phenotype data, but already ran the location analysis (GWAS) using other methods (e.g. PLINK or BOLT-LMM) or plan to use published GWAS p-values. I am interested in the "</em><em>gS</em><em>" analysis and the combined "</em><em>gJLS</em><em>" analyses using existing location p-values in place of the gL.</em> <span style="color:green">The user can run scale analysis alone using the preferred approach depending on sample size and computational requirements. See next scenario for combining gL and gS. </span></li>
</ul>
<h2 id="gjls-with-summary-level-genotype-and-phenotype-data"><strong>gJLS</strong> with <strong>SUMMARY-LEVEL</strong> genotype and phenotype data.</h2>
<p><em>I only have "</em><em>SUMMARY-LEVEL</em><em>" data (e.g. p-values) for location and scale association from external sources (e.g. GIANT) and am interested in the "</em><em>gJLS</em><em>" analysis.</em> <span style="color:green">Once the gL and gS <em>p</em>-values have been obtained, the function <em>gJLS2s()</em> can be called in R GUI or computed via the Rscript to produce the combined gJLS <em>p</em>-values without access to individual-level data.</span></p>
              
            </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
      
        <a href="../nonGUI/" class="btn btn-neutral" title="nonGUI"><span class="icon icon-circle-arrow-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
      
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
    
      <span><a href="../nonGUI/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
  </span>
</div>
    <script>var base_url = '..';</script>
    <script src="../js/theme_extra.js" defer></script>
    <script src="../js/theme.js" defer></script>
      <script src="../search/main.js" defer></script>
    <script defer>
        window.onload = function () {
            SphinxRtdTheme.Navigation.enable(true);
        };
    </script>

</body>
</html>
